src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DocumentMapping/RazorEditHelper.TextChangeBuilder.cs

```diff
@@ -0,0 +1,427 @@
+            var csharpDocument = codeDocument.GetCSharpDocument();
+            var csharpText = csharpDocument.GetGeneratedSourceText();
+
+            foreach (var edit in csharpEdits)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var linePositionSpan = csharpText.GetLinePositionSpan(edit.Span.ToTextSpan());
+
+                if (!_documentMappingService.TryMapToHostDocumentRange(
+                    csharpDocument,
+                    linePositionSpan,
+                    MappingBehavior.Strict,
+                    out var mappedLinePositionSpan))
+                {
+                    continue;
+                }
+
+                var mappedSpan = razorText.GetTextSpan(mappedLinePositionSpan);
+                var node = root.FindNode(mappedSpan, getInnermostNodeForTie: true);
+                if (node is null)
+                {
+                    continue;
+                }
+
+                if (RazorSyntaxFacts.IsInUsingDirective(node))
+                {
+                    continue;
+                }
+
+                var mappedEdit = new RazorTextChange()
+                {
+                    Span = mappedSpan.ToRazorTextSpan(),
+                    NewText = edit.NewText
+                };
+                _builder.Add(mappedEdit);
+            }
+        }
+
+        /// <summary>
+        /// Given a set of new and removed usings, adds text changes to this builder using the following logic:
+        ///
+        /// <list type="number">
+        /// <item>
+        /// If there are no existing usings the new usings are added at the top of the document following any page, component, or namespace directives.
+        /// </item>
+        ///
+        /// <item>
+        /// If there are existing usings but they are in a continous block, replace that block with the new ordered set of usings.
+        /// </item>
+        ///
+        /// <item>
+        /// If for some reason a user has usings not in a single block (allows for whitespace), then replace the first block of using directives
+        /// with the set of usings within that block that have not been removed AND the new usings. The remaining directives outside the block are removed
+        /// as needed.
+        /// </item>
+        /// </list>
+        /// </summary>
+        public void AddUsingsChanges(
+            RazorCodeDocument codeDocument,
+            ImmutableArray<string> addedUsings,
+            ImmutableArray<string> removedUsings,
+            CancellationToken cancellationToken)
+        {
+            if (addedUsings.Length == 0 && removedUsings.Length == 0)
+            {
+                return;
+            }
+
+            // If only usings were added then just need to find where to insert them.
+            if (removedUsings.Length == 0)
+            {
+                AddNewUsingChanges(codeDocument, addedUsings, cancellationToken);
```

Comment:
Nit: The AddComplex and AddRemove methods have the plural "Using**s**" in their name